--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,7 @@
 cmake_minimum_required (VERSION 3.9)
 project (nix-plugins)
+set (CMAKE_CXX_STANDARD 23)
+set (CMAKE_CXX_STANDARD_REQUIRED ON)
 set (nix-plugins_VERSION_MAJOR 16)
 set (nix-plugins_VERSION_MINOR 0)
 set (nix-plugins_VERSION_PATCH 0)
@@ -5,7 +7,7 @@ set (nix-plugins_VERSION_PATCH 0)
 
 find_package(PkgConfig)
 
-pkg_check_modules(NIX REQUIRED nix-cmd>=2.28 nix-expr>=2.28 nix-main>=2.28 nix-store>=2.28)
+pkg_check_modules(NIX REQUIRED lix-cmd lix-expr lix-main lix-store)
 
 find_path(BOOST_INCLUDE_DIR boost/format.hpp)
 if(BOOST_INCLUDE_DIR STREQUAL "BOOST_INCLUDE_DIR-NOTFOUND")
--- a/extra-builtins.cc
+++ b/extra-builtins.cc
@@ -1,11 +1,12 @@
-#include <nix/cmd/common-eval-args.hh>
-#include <nix/expr/eval-settings.hh>
-#include <nix/expr/primops.hh>
-#include <nix/fetchers/filtering-source-accessor.hh>
-#include <nix/store/globals.hh>
-#include <nix/util/configuration.hh>
-#include <nix/util/config-global.hh>
-
+#include <lix/config.h>
+#include <lix/libcmd/common-eval-args.hh>
+#include <lix/libexpr/eval-inline.hh>
+#include <lix/libexpr/eval-settings.hh>
+#include <lix/libexpr/primops.hh>
+#include <lix/libstore/globals.hh>
+#include <lix/libutil/canon-path.hh>
+#include <lix/libutil/config.hh>
+#include <cerrno>
 
 #include "nix-plugins-config.h"
 
@@ -22,41 +23,42 @@
 
 static GlobalConfig::Register rp(&extraBuiltinsSettings);
 
-static void extraBuiltins(EvalState & state, const PosIdx pos,
+static void extraBuiltins(EvalState & state,
     Value ** _args, Value & v)
 {
-    static auto extraBuiltinsFile = state.rootPath(CanonPath(extraBuiltinsSettings.extraBuiltinsFile.to_string()));
-    if (auto rootFS2 = state.rootFS.dynamic_pointer_cast<AllowListSourceAccessor>())
-        rootFS2->allowPrefix(CanonPath(extraBuiltinsFile.path.abs()));
+    static auto extraBuiltinsFile = SourcePath(CanonPath(extraBuiltinsSettings.extraBuiltinsFile.to_string()));
+    state.ctx.paths.allowPath(extraBuiltinsFile.to_string());
 
     try {
-        auto fun = state.allocValue();
-        state.evalFile(extraBuiltinsFile, *fun);
-        Value * arg;
+        Value fun;
+        state.evalFile(extraBuiltinsFile, fun);
+        Value arg;
         if (evalSettings.enableNativeCode) {
-            arg = state.baseEnv.values[0];
+            arg = state.ctx.builtins.get("builtins");
         } else {
-            auto attrs = state.buildBindings(2);
+            auto attrs = state.ctx.buildBindings(2);
 
-            auto sExec = state.symbols.create("exec");
-            attrs.alloc(sExec).mkPrimOp(new PrimOp {
+            auto sExec = state.ctx.symbols.create("exec");
+            attrs.alloc(sExec).mkPrimOp(new PrimOp(PrimOpDetails{
                 .name = "exec",
                 .arity = 1,
                 .fun = prim_exec,
-            });
+            }));
 
-            auto sImportNative = state.symbols.create("importNative");
-            attrs.alloc(sImportNative).mkPrimOp(new PrimOp {
+            auto sImportNative = state.ctx.symbols.create("importNative");
+            attrs.alloc(sImportNative).mkPrimOp(new PrimOp(PrimOpDetails{
                 .name = "importNative",
                 .arity = 2,
                 .fun = prim_importNative,
-            });
+            }));
 
-            arg = state.allocValue();
-            arg->mkAttrs(attrs);
+            arg.mkAttrs(attrs);
         }
-        v.mkApp(fun, arg);
-        state.forceValue(v, pos);
+        v = {NewValueAs::app, state.ctx.mem, fun, arg};
+        state.forceValue(v, noPos);
-    } catch (FileNotFound &) {
-        v.mkNull();
+    } catch (SysError & e) {
+        if (e.errNo == ENOENT)
+            v.mkNull();
+        else
+            throw;
     }
@@ -63,23 +65,23 @@
 }
 
-static RegisterPrimOp rp1({
+static RegisterPrimOp rp1(PrimOpDetails{
     .name = "__extraBuiltins",
     .arity = 0,
     .fun = extraBuiltins,
 });
 
-static void cflags(EvalState & state, const PosIdx _pos,
+static void cflags(EvalState & state,
     Value ** _args, Value & v)
 {
-    auto attrs = state.buildBindings(3);
+    auto attrs = state.ctx.buildBindings(3);
     attrs.alloc("NIX_INCLUDE_DIRS").mkString(NIX_INCLUDE_DIRS);
     attrs.alloc("NIX_CFLAGS_OTHER").mkString(NIX_CFLAGS_OTHER);
     attrs.alloc("BOOST_INCLUDE_DIR").mkString(BOOST_INCLUDE_DIR);
     v.mkAttrs(attrs);
 }
 
-static RegisterPrimOp rp2({
+static RegisterPrimOp rp2(PrimOpDetails{
     .name = "__nix-cflags",
     .arity = 0,
     .fun = cflags,
 });
